File: frontend_flutter/lib/features/model/employee.model.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:logger/logger.dart';

final logger = Logger();

/// PTInfo - Thông tin chi tiết dành cho Personal Trainer
class PTInfo {
  List<String> specialties; // Chuyên môn
  int experience; // Số năm kinh nghiệm
  double rating; // Đánh giá trung bình (0-5)
  int totalRatings; // Tổng số lượt đánh giá
  String bio; // Mô tả bản thân
  List<dynamic> certificates; // Chứng chỉ (có thể là String hoặc Object)
  List<String> availableHours; // Khung giờ làm việc
  int maxClientsPerDay; // Số khách hàng tối đa/ngày
  bool isAcceptingNewClients; // Có nhận khách mới không
  List<String> languages; // Ngôn ngữ
  List<dynamic> achievements; // Thành tích (có thể là String hoặc Object)
  Map<String, String> socialMedia; // Mạng xã hội

  PTInfo({
    this.specialties = const [],
    this.experience = 0,
    this.rating = 0.0,
    this.totalRatings = 0,
    this.bio = '',
    this.certificates = const [],
    this.availableHours = const [],
    this.maxClientsPerDay = 8,
    this.isAcceptingNewClients = true,
    this.languages = const ['vi'],
    this.achievements = const [],
    this.socialMedia = const {},
  });

  factory PTInfo.fromMap(Map<String, dynamic> data) {
    return PTInfo(
      specialties: List<String>.from(data['specialties'] ?? []),
      experience: data['experience'] ?? 0,
      rating: (data['rating'] ?? 0).toDouble(),
      totalRatings: data['totalRatings'] ?? 0,
      bio: data['bio'] ?? '',
      certificates: List<dynamic>.from(data['certificates'] ?? []),
      availableHours: List<String>.from(data['availableHours'] ?? []),
      maxClientsPerDay: data['maxClientsPerDay'] ?? 8,
      isAcceptingNewClients: data['isAcceptingNewClients'] ?? true,
      languages: List<String>.from(data['languages'] ?? ['vi']),
      achievements: List<dynamic>.from(data['achievements'] ?? []),
      socialMedia: Map<String, String>.from(data['socialMedia'] ?? {}),
    );
  }

  Map<String, dynamic> toMap() {
    return {
      'specialties': specialties,
      'experience': experience,
      'rating': rating,
      'totalRatings': totalRatings,
      'bio': bio,
      'certificates': certificates,
      'availableHours': availableHours,
      'maxClientsPerDay': maxClientsPerDay,
      'isAcceptingNewClients': isAcceptingNewClients,
      'languages': languages,
      'achievements': achievements,
      'socialMedia': socialMedia,
    };
  }

  /// Get formatted rating text (e.g., "4.5 ⭐")
  String get formattedRating {
    return '${rating.toStringAsFixed(1)} ⭐';
  }

  /// Get experience text (e.g., "5 năm kinh nghiệm")
  String get experienceText {
    if (experience == 0) return 'Mới vào nghề';
    if (experience == 1) return '1 năm kinh nghiệm';
    return '$experience năm kinh nghiệm';
  }

  /// Check if has high rating (>= 4.5)
  bool get hasHighRating => rating >= 4.5;

  /// Check if experienced (>= 3 years)
  bool get isExperienced => experience >= 3;
}

/// EmployeeModel - Model nhân viên phía client
class EmployeeModel {
  String id; // Firestore document ID
  String fullName;
  String gender; // 'male', 'female', 'other'
  DateTime? dateOfBirth;
  String phone;
  String email;
  String address;
  String
  position; // 'PT', 'Lễ tân', 'Quản lý', 'Kế toán', 'Bảo vệ', 'Vệ sinh', 'Khác'
  DateTime? startDate;
  String status; // 'active', 'inactive', 'resigned', 'suspended'
  String shift; // 'morning', 'afternoon', 'evening', 'full-time', 'part-time'
  String? uid; // Firebase Auth UID
  String role; // 'employee', 'pt', 'manager', 'admin'
  DateTime? createdAt;
  DateTime? updatedAt;
  double salary;
  double commissionRate; // Tỷ lệ hoa hồng (%)
  int totalClients; // Tổng số khách hàng (dùng cho PT)
  String avatarUrl;
  String idCard; // CMND/CCCD
  String notes;
  PTInfo? ptInfo; // Thông tin PT (chỉ có khi position = 'PT')

  EmployeeModel({
    this.id = '',
    this.fullName = '',
    this.gender = 'male',
    this.dateOfBirth,
    this.phone = '',
    this.email = '',
    this.address = '',
    this.position = '',
    this.startDate,
    this.status = 'active',
    this.shift = '',
    this.uid,
    this.role = 'employee',
    this.createdAt,
    this.updatedAt,
    this.salary = 0.0,
    this.commissionRate = 0.0,
    this.totalClients = 0,
    this.avatarUrl = '',
    this.idCard = '',
    this.notes = '',
    this.ptInfo,
  });

  /// Chuyển đổi từ Firestore DocumentSnapshot
  factory EmployeeModel.fromFirestore(DocumentSnapshot doc) {
    final data = doc.data() as Map<String, dynamic>;
    return EmployeeModel.fromMap(data, doc.id);
  }

  /// Chuyển đổi từ Map
  factory EmployeeModel.fromMap(Map<String, dynamic> data, String docId) {
    return EmployeeModel(
      id: docId,
      fullName: data['fullName'] ?? '',
      gender: data['gender'] ?? 'male',
      dateOfBirth: _parseTimestamp(data['dateOfBirth']),
      phone: data['phone'] ?? '',
      email: data['email'] ?? '',
      address: data['address'] ?? '',
      position: data['position'] ?? '',
      startDate: _parseTimestamp(data['startDate']),
      status: data['status'] ?? 'active',
      shift: data['shift'] ?? '',
      uid: data['uid'],
      role: data['role'] ?? 'employee',
      createdAt: _parseTimestamp(data['createdAt']),
      updatedAt: _parseTimestamp(data['updatedAt']),
      salary: _parseDouble(data['salary']) ?? 0.0,
      commissionRate: _parseDouble(data['commissionRate']) ?? 0.0,
      totalClients: data['totalClients'] ?? 0,
      avatarUrl: data['avatarUrl'] ?? '',
      idCard: data['idCard'] ?? '',
      notes: data['notes'] ?? '',
      ptInfo: data['ptInfo'] != null
          ? PTInfo.fromMap(data['ptInfo'] as Map<String, dynamic>)
          : null,
    );
  }

  /// Chuyển đổi sang Map để lưu vào Firestore
  Map<String, dynamic> toMap() {
    final map = {
      'fullName': fullName,
      'gender': gender,
      'dateOfBirth': dateOfBirth?.toIso8601String(),
      'phone': phone,
      'email': email,
      'address': address,
      'position': position,
      'startDate': startDate?.toIso8601String(),
      'status': status,
      'shift': shift,
      'uid': uid,
      'role': role,
      'createdAt': createdAt?.toIso8601String(),
      'updatedAt': updatedAt?.toIso8601String(),
      'salary': salary,
      'commissionRate': commissionRate,
      'totalClients': totalClients,
      'avatarUrl': avatarUrl,
      'idCard': idCard,
      'notes': notes,
    };

    // Chỉ thêm ptInfo nếu là PT
    if (position == 'PT' && ptInfo != null) {
      map['ptInfo'] = ptInfo!.toMap();
    }

    return map;
  }

  /// Helper: Parse Timestamp từ Firestore
  static DateTime? _parseTimestamp(dynamic value) {
    if (value == null) return null;
    if (value is Timestamp) return value.toDate();
    if (value is DateTime) return value;
    if (value is String) {
      try {
        return DateTime.parse(value);
      } catch (e) {
        logger.e('Không thể parse DateTime từ string: $value');
        return null;
      }
    }
    return null;
  }

  /// Helper: Parse double
  static double? _parseDouble(dynamic value) {
    if (value == null) return null;
    if (value is double) return value;
    if (value is int) return value.toDouble();
    if (value is String) {
      try {
        return double.parse(value);
      } catch (e) {
        logger.e('Không thể parse double từ string: $value');
        return null;
      }
    }
    return null;
  }

  // ==================== COMPUTED PROPERTIES ====================

  /// Check if employee is PT
  bool get isPT => position == 'PT';

  /// Check if active
  bool get isActive => status == 'active';

  /// Get age from date of birth
  int? get age {
    if (dateOfBirth == null) return null;
    final now = DateTime.now();
    int age = now.year - dateOfBirth!.year;
    if (now.month < dateOfBirth!.month ||
        (now.month == dateOfBirth!.month && now.day < dateOfBirth!.day)) {
      age--;
    }
    return age;
  }

  /// Get years of service
  int? get yearsOfService {
    if (startDate == null) return null;
    final now = DateTime.now();
    int years = now.year - startDate!.year;
    if (now.month < startDate!.month ||
        (now.month == startDate!.month && now.day < startDate!.day)) {
      years--;
    }
    return years < 0 ? 0 : years;
  }

  /// Get status display text in Vietnamese
  String get statusText {
    switch (status) {
      case 'active':
        return 'Đang làm việc';
      case 'inactive':
        return 'Tạm nghỉ';
      case 'resigned':
        return 'Đã nghỉ việc';
      case 'suspended':
        return 'Bị đình chỉ';
      default:
        return 'Không xác định';
    }
  }

  /// Get shift display text in Vietnamese
  String get shiftText {
    switch (shift) {
      case 'morning':
        return 'Ca sáng';
      case 'afternoon':
        return 'Ca chiều';
      case 'evening':
        return 'Ca tối';
      case 'full-time':
        return 'Toàn thời gian';
      case 'part-time':
        return 'Bán thời gian';
      default:
        return 'Chưa xác định';
    }
  }

  /// Get gender display text in Vietnamese
  String get genderText {
    switch (gender) {
      case 'male':
        return 'Nam';
      case 'female':
        return 'Nữ';
      case 'other':
        return 'Khác';
      default:
        return 'Chưa xác định';
    }
  }

  // ==================== STATIC METHODS ====================

  /// Lấy nhân viên theo ID
  static Future<EmployeeModel?> getById(String id) async {
    try {
      final doc = await FirebaseFirestore.instance
          .collection('employees')
          .doc(id)
          .get();
      if (doc.exists) {
        return EmployeeModel.fromFirestore(doc);
      }
      return null;
    } catch (e) {
      logger.e('Error getting employee by ID: $e');
      return null;
    }
  }

  /// Lấy tất cả PT (Personal Trainers)
  static Future<List<EmployeeModel>> getAllPTs({
    bool onlyActive = true,
    bool onlyAcceptingClients = false,
  }) async {
    try {
      Query query = FirebaseFirestore.instance
          .collection('employees')
          .where('position', isEqualTo: 'PT');

      if (onlyActive) {
        query = query.where('status', isEqualTo: 'active');
      }

      final querySnapshot = await query.get();

      List<EmployeeModel> pts = querySnapshot.docs
          .map((doc) => EmployeeModel.fromFirestore(doc))
          .toList();

      // Filter by accepting clients (client-side vì không thể query nested field)
      if (onlyAcceptingClients) {
        pts = pts
            .where((pt) => pt.ptInfo?.isAcceptingNewClients ?? false)
            .toList();
      }

      // Sort by rating
      pts.sort((a, b) {
        final aRating = a.ptInfo?.rating ?? 0;
        final bRating = b.ptInfo?.rating ?? 0;
        return bRating.compareTo(aRating);
      });
      logger.i('Fetched ${pts.length} PTs from Firestore');

      return pts;
    } catch (e) {
      logger.e('Error getting all PTs: $e');
      return [];
    }
  }

  /// Lấy PTs theo chuyên môn
  static Future<List<EmployeeModel>> getPTsBySpecialty(
    String specialty, {
    bool onlyActive = true,
  }) async {
    try {
      final allPTs = await getAllPTs(onlyActive: onlyActive);

      // Filter by specialty (client-side filtering)
      final filteredPTs = allPTs.where((pt) {
        return pt.ptInfo?.specialties.contains(specialty) ?? false;
      }).toList();

      return filteredPTs;
    } catch (e) {
      logger.e('Error getting PTs by specialty: $e');
      return [];
    }
  }

  /// Lấy top rated PTs
  static Future<List<EmployeeModel>> getTopRatedPTs({int limit = 5}) async {
    try {
      final allPTs = await getAllPTs(onlyActive: true);

      // Sort by rating and total ratings
      allPTs.sort((a, b) {
        final aRating = a.ptInfo?.rating ?? 0;
        final bRating = b.ptInfo?.rating ?? 0;
        final aTotal = a.ptInfo?.totalRatings ?? 0;
        final bTotal = b.ptInfo?.totalRatings ?? 0;

        // Primary sort: rating
        final ratingCompare = bRating.compareTo(aRating);
        if (ratingCompare != 0) return ratingCompare;

        // Secondary sort: total ratings
        return bTotal.compareTo(aTotal);
      });

      return allPTs.take(limit).toList();
    } catch (e) {
      logger.e('Error getting top rated PTs: $e');
      return [];
    }
  }

  /// Search employees
  static Future<List<EmployeeModel>> searchEmployees(
    String searchQuery, {
    String? position,
    String? status,
    int limit = 20,
  }) async {
    try {
      Query query = FirebaseFirestore.instance.collection('employees');

      if (position != null && position.isNotEmpty) {
        query = query.where('position', isEqualTo: position);
      }

      if (status != null && status.isNotEmpty) {
        query = query.where('status', isEqualTo: status);
      }

      query = query.limit(limit);

      final querySnapshot = await query.get();
      List<EmployeeModel> employees = querySnapshot.docs
          .map((doc) => EmployeeModel.fromFirestore(doc))
          .toList();

      // Client-side search filtering
      if (searchQuery.isNotEmpty) {
        final lowerQuery = searchQuery.toLowerCase();
        employees = employees.where((emp) {
          return emp.fullName.toLowerCase().contains(lowerQuery) ||
              emp.email.toLowerCase().contains(lowerQuery) ||
              emp.phone.contains(lowerQuery);
        }).toList();
      }

      return employees;
    } catch (e) {
      logger.e('Error searching employees: $e');
      return [];
    }
  }

  /// Get employee statistics
  static Future<Map<String, int>> getStatistics() async {
    try {
      final querySnapshot = await FirebaseFirestore.instance
          .collection('employees')
          .get();

      int total = querySnapshot.docs.length;
      int active = 0;
      int pts = 0;
      int inactive = 0;

      for (var doc in querySnapshot.docs) {
        final data = doc.data();
        if (data['status'] == 'active') active++;
        if (data['position'] == 'PT') pts++;
        if (data['status'] == 'inactive' || data['status'] == 'resigned') {
          inactive++;
        }
      }

      return {
        'total': total,
        'active': active,
        'pts': pts,
        'inactive': inactive,
      };
    } catch (e) {
      logger.e('Error getting employee statistics: $e');
      return {'total': 0, 'active': 0, 'pts': 0, 'inactive': 0};
    }
  }
}

File: frontend_react/src/firebase/lib/features/employee/employee.model.js
import { 
  collection, 
  addDoc, 
  updateDoc, 
  deleteDoc, 
  doc, 
  getDocs, 
  getDoc,
  getCountFromServer,
  query, 
  where, 
  orderBy, 
  limit as fsLimit, 
  startAfter as fsStartAfter,
  Timestamp 
} from 'firebase/firestore';
import { db } from '../../config/firebase.js';
import Joi from 'joi';

export class EmployeeModel {
  constructor({
    _id = null,
    fullName = '',
    gender = 'male',
    dateOfBirth = null,
    phone = '',
    email = '',
    address = '',
    position = '',
    startDate = null,
    status = 'active',
    shift = '',
    uid = null,
    role = 'employee',
    createdAt = new Date(),
    updatedAt = new Date(),
    salary = 0,
    commissionRate = 0,
    totalClients = 0,
    avatarUrl = '',
    idCard = '',
    notes = '',
    ptInfo = null,
    faceRegistered = false
  } = {}) {
    this._id = _id;
    this.fullName = fullName;
    this.gender = gender;
    this.dateOfBirth = dateOfBirth;
    this.phone = phone;
    this.email = email;
    this.address = address;
    this.position = position;
    this.startDate = startDate;
    this.status = status;
    this.shift = shift;
    this.uid = uid;
    this.role = role;
    this.createdAt = createdAt;
    this.updatedAt = updatedAt;
    this.salary = salary;
    this.commissionRate = commissionRate;
    this.totalClients = totalClients;
    this.avatarUrl = avatarUrl;
    this.idCard = idCard;
    this.notes = notes;
    this.faceRegistered = faceRegistered;
    
    // PT-specific information (only for PT position)
    this.ptInfo = ptInfo || (position === 'PT' ? {
      specialties: [], // ["Giảm cân", "Tăng cơ", "Yoga", "Cardio", "Phục hồi chấn thương"]
      experience: 0, // Số năm kinh nghiệm
      rating: 0, // Đánh giá trung bình (0-5)
      totalRatings: 0, // Tổng số lượt đánh giá
      bio: '', // Mô tả bản thân PT
      certificates: [], // Các chứng chỉ ["CPT", "Yoga Instructor", "Nutrition Specialist"]
      availableHours: [], // Khung giờ có thể làm việc ["06:00-08:00", "18:00-20:00"]
      maxClientsPerDay: 8, // Số lượng khách hàng tối đa trong ngày
      isAcceptingNewClients: true, // Có nhận khách hàng mới không
      languages: ['vi'], // Ngôn ngữ ["vi", "en"]
      achievements: [], // Thành tích ["Trainer of the Year 2023"]
      socialMedia: { // Mạng xã hội
        facebook: '',
        instagram: '',
        youtube: '',
        tiktok: ''
      }
    } : null);
  }

  // Validation schema
  static validationSchema = Joi.object({
    fullName: Joi.string().min(2).max(100).required().messages({
      'string.empty': 'Họ tên không được để trống',
      'string.min': 'Họ tên phải có ít nhất 2 ký tự',
      'string.max': 'Họ tên không được quá 100 ký tự'
    }),
    gender: Joi.string().valid('male', 'female', 'other').required().messages({
      'any.only': 'Giới tính phải là male, female hoặc other'
    }),
    dateOfBirth: Joi.date().max('now').required().messages({
      'date.max': 'Ngày sinh không được lớn hơn ngày hiện tại'
    }),
    phone: Joi.string().pattern(/^[0-9]{10,11}$/).required().messages({
      'string.pattern.base': 'Số điện thoại phải có 10-11 chữ số'
    }),
    email: Joi.string().email().required().messages({
      'string.email': 'Email không hợp lệ'
    }),
    address: Joi.string().max(200).required().messages({
      'string.max': 'Địa chỉ không được quá 200 ký tự'
    }),
    position: Joi.string().valid('PT', 'Lễ tân', 'Quản lý', 'Kế toán', 'Bảo vệ', 'Vệ sinh', 'Khác').required().messages({
      'any.only': 'Vị trí làm việc không hợp lệ'
    }),
    startDate: Joi.date().max('now').required().messages({
      'date.max': 'Ngày bắt đầu không được lớn hơn ngày hiện tại'
    }),
    status: Joi.string().valid('active', 'inactive', 'resigned', 'suspended').default('active'),
    shift: Joi.string().valid('fulltime', 'parttime').required(),
    role: Joi.string().valid('employee', 'pt', 'manager', 'admin').default('employee'),
    salary: Joi.number().min(0).required().messages({
      'number.min': 'Lương không được âm'
    }),
    commissionRate: Joi.number().min(0).max(100).default(0).messages({
      'number.min': 'Tỷ lệ hoa hồng không được âm',
      'number.max': 'Tỷ lệ hoa hồng không được quá 100%'
    }),
    totalClients: Joi.number().min(0).default(0),
    avatarUrl: Joi.string().allow('', null).optional(),
    idCard: Joi.string().pattern(/^(?:\d{9}|\d{12})$/).required().messages({
      'string.pattern.base': 'Căn cước công dân phải có 9 hoặc 12 chữ số'
    }),
    notes: Joi.string().max(500).allow('').optional().messages({
      'string.max': 'Ghi chú không được quá 500 ký tự'
    }),
    uid: Joi.string().allow(null).optional(),
    faceRegistered: Joi.boolean().default(false).optional(),
    
    // PT Info validation (chỉ validate khi position = 'PT')
    ptInfo: Joi.when('position', {
      is: 'PT',
      then: Joi.object({
        specialties: Joi.array().items(
          Joi.string().valid('Giảm cân', 'Tăng cơ', 'Yoga', 'Cardio', 'Phục hồi chấn thương', 'Powerlifting', 'CrossFit', 'Pilates', 'Boxing', 'Khác')
        ).max(5).optional(),
        experience: Joi.number().min(0).max(50).optional().messages({
          'number.min': 'Kinh nghiệm không được âm',
          'number.max': 'Kinh nghiệm không được quá 50 năm'
        }),
        rating: Joi.number().min(0).max(5).optional(),
        totalRatings: Joi.number().min(0).optional(),
        bio: Joi.string().max(1000).allow('').optional().messages({
          'string.max': 'Mô tả không được quá 1000 ký tự'
        }),
        certificates: Joi.array().items(
          Joi.alternatives().try(
            Joi.string().max(100), // Legacy format
            Joi.object({
              id: Joi.string().required(),
              text: Joi.string().max(100).required(),
              images: Joi.array().items(
                Joi.object({
                  id: Joi.string().required(),
                  url: Joi.string().required(),
                  path: Joi.string().optional(),
                  fileName: Joi.string().optional(),
                  name: Joi.string().optional(),
                  size: Joi.number().optional(),
                  uploadedAt: Joi.date().optional()
                })
              ).optional().default([])
            }).required() // New format
          )
        ).max(10).optional(),
        availableHours: Joi.array().items(
          Joi.string().pattern(/^\d{2}:\d{2}-\d{2}:\d{2}$/)
        ).max(10).optional().messages({
          'string.pattern.base': 'Khung giờ phải có định dạng HH:MM-HH:MM'
        }),
        maxClientsPerDay: Joi.number().min(1).max(20).default(8).optional(),
        isAcceptingNewClients: Joi.boolean().default(true).optional(),
        languages: Joi.array().items(
          Joi.string().valid('vi', 'en', 'ja', 'ko', 'zh')
        ).default(['vi']).optional(),
        achievements: Joi.array().items(
          Joi.alternatives().try(
            Joi.string().max(200), // Legacy format  
            Joi.object({
              id: Joi.string().required(),
              text: Joi.string().max(200).required(),
              images: Joi.array().items(
                Joi.object({
                  id: Joi.string().required(),
                  url: Joi.string().required(),
                  path: Joi.string().optional(),
                  fileName: Joi.string().optional(),
                  name: Joi.string().optional(),
                  size: Joi.number().optional(),
                  uploadedAt: Joi.date().optional()
                })
              ).optional().default([])
            }).required() // New format
          )
        ).max(20).optional(),
        socialMedia: Joi.object({
          facebook: Joi.string().allow('').optional(),
          instagram: Joi.string().allow('').optional(),
          youtube: Joi.string().allow('').optional(),
          tiktok: Joi.string().allow('').optional()
        }).optional()
      }).optional(),
      otherwise: Joi.optional()
    })
  });

  // Normalize shift values to new format
  static normalizeShift(shift) {
    const shiftMap = {
      'full-time': 'fulltime',
      'part-time': 'parttime',
      'morning': 'parttime',
      'afternoon': 'parttime', 
      'evening': 'parttime'
    };
    return shiftMap[shift] || shift;
  }

  // Validate employee data
  validate() {
    // Normalize shift before validation
    const normalizedShift = EmployeeModel.normalizeShift(this.shift);
    
    const { error } = EmployeeModel.validationSchema.validate({
      fullName: this.fullName,
      gender: this.gender,
      dateOfBirth: this.dateOfBirth,
      phone: this.phone,
      email: this.email,
      address: this.address,
      position: this.position,
      startDate: this.startDate,
      status: this.status,
      shift: normalizedShift,
      role: this.role,
      salary: this.salary,
      commissionRate: this.commissionRate,
      totalClients: this.totalClients,
      avatarUrl: this.avatarUrl,
      idCard: this.idCard,
      notes: this.notes,
      uid: this.uid,
      ptInfo: this.ptInfo
    });
    
    if (error) {
      throw new Error(error.details[0].message);
    }
    
    // Update shift to normalized value
    this.shift = normalizedShift;
    return true;
  }

  // Convert to Firestore format
  toFirestore() {
    const data = {
      fullName: this.fullName,
      gender: this.gender,
      dateOfBirth: this.dateOfBirth ? Timestamp.fromDate(new Date(this.dateOfBirth)) : null,
      phone: this.phone,
      email: this.email,
      address: this.address,
      position: this.position,
      startDate: this.startDate ? Timestamp.fromDate(new Date(this.startDate)) : null,
      status: this.status,
      shift: this.shift,
      uid: this.uid,
      role: this.role,
      createdAt: this.createdAt ? Timestamp.fromDate(new Date(this.createdAt)) : Timestamp.now(),
      updatedAt: Timestamp.now(),
      salary: this.salary,
      commissionRate: this.commissionRate,
      totalClients: this.totalClients,
      avatarUrl: this.avatarUrl || '',
      idCard: this.idCard || '',
      notes: this.notes || '',
      faceRegistered: this.faceRegistered || false
    };

    // Chỉ thêm ptInfo nếu là PT
    if (this.position === 'PT' && this.ptInfo) {
      data.ptInfo = this.ptInfo;
    }

    // Remove null values
    Object.keys(data).forEach(key => {
      if (data[key] === null || data[key] === undefined) {
        delete data[key];
      }
    });

    return data;
  }

  // Create new employee
  async save() {
    try {
      this.validate();
      const employeesCollectionRef = collection(db, 'employees');
      
      // Check if email already exists
      const emailQuery = query(employeesCollectionRef, where('email', '==', this.email));
      const emailSnapshot = await getDocs(emailQuery);
      
      if (!emailSnapshot.empty && (!this._id || emailSnapshot.docs[0].id !== this._id)) {
        throw new Error('Email đã tồn tại trong hệ thống');
      }

      // Check if phone already exists  
      const phoneQuery = query(employeesCollectionRef, where('phone', '==', this.phone));
      const phoneSnapshot = await getDocs(phoneQuery);
      
      if (!phoneSnapshot.empty && (!this._id || phoneSnapshot.docs[0].id !== this._id)) {
        throw new Error('Số điện thoại đã tồn tại trong hệ thống');
      }

      if (this._id) {
        // Update existing employee
        const employeeDocRef = doc(db, 'employees', this._id);
        const updateData = this.toFirestore();
        await updateDoc(employeeDocRef, updateData);
        return this._id;
      } else {
        // Create new employee
        const docRef = await addDoc(employeesCollectionRef, this.toFirestore());
        this._id = docRef.id;
        return docRef.id;
      }
    } catch (error) {
      console.error('Error saving employee:', error);
      throw error;
    }
  }

  // Get employee by ID
  static async getById(id) {
    try {
      const employeeDocRef = doc(db, 'employees', id);
      const docSnap = await getDoc(employeeDocRef);
      
      if (docSnap.exists()) {
        const data = docSnap.data();
        
        // Convert Timestamps back to dates
        Object.keys(data).forEach(field => {
          if (data[field] instanceof Timestamp) {
            data[field] = data[field].toDate();
          }
        });
        
        return new EmployeeModel({ _id: docSnap.id, ...data });
      } else {
        throw new Error('Không tìm thấy nhân viên');
      }
    } catch (error) {
      console.error('Error getting employee:', error);
      throw error;
    }
  }

  // Get all employees with filters and pagination
  static async getAll(filters = {}, limit = 10, startAfterDoc = null) {
    try {
      const employeesCollectionRef = collection(db, 'employees');
      const queryConstraints = [];

      // Add filters
      if (filters.status) {
        queryConstraints.push(where('status', '==', filters.status));
      }

      if (filters.position) {
        queryConstraints.push(where('position', '==', filters.position));
      }

      if (filters.role) {
        queryConstraints.push(where('role', '==', filters.role));
      }

      if (filters.searchQuery && !filters.status && !filters.position && !filters.role) {
        // For search, we'll get all and filter client-side (similar to user search)
        const effectiveLimit = Math.max(limit, 50);
        queryConstraints.push(orderBy('fullName'));
        queryConstraints.push(fsLimit(effectiveLimit));
      } else {
        queryConstraints.push(orderBy('createdAt', 'desc'));
        queryConstraints.push(fsLimit(limit));
      }

      if (startAfterDoc) {
        queryConstraints.push(fsStartAfter(startAfterDoc));
      }

      const finalQuery = query(employeesCollectionRef, ...queryConstraints);
      const querySnapshot = await getDocs(finalQuery);

      let employees = querySnapshot.docs.map(docSnap => {
        const data = docSnap.data();
        
        // Convert Timestamps to dates
        Object.keys(data).forEach(field => {
          if (data[field] instanceof Timestamp) {
            data[field] = data[field].toDate();
          }
        });
        
        return new EmployeeModel({ _id: docSnap.id, ...data });
      });

      // Client-side search filter
      if (filters.searchQuery && !filters.status && !filters.position && !filters.role) {
        const searchQuery = filters.searchQuery.trim().toLowerCase();
        employees = employees.filter(emp => 
          (emp.fullName && emp.fullName.toLowerCase().includes(searchQuery)) ||
          (emp.email && emp.email.toLowerCase().includes(searchQuery)) ||
          (emp.phone && emp.phone.includes(searchQuery))
        );
        employees = employees.slice(0, limit);
      }

      const lastDoc = querySnapshot.docs.length > 0
        ? querySnapshot.docs[querySnapshot.docs.length - 1]
        : null;

      const hasMore = filters.searchQuery 
        ? employees.length === limit
        : querySnapshot.docs.length === limit;

      return { employees, lastDoc, hasMore };
    } catch (error) {
      console.error('Error getting employees:', error);
      throw error;
    }
  }

  // Delete employee
  static async delete(id) {
    try {
      const employeeDocRef = doc(db, 'employees', id);
      await deleteDoc(employeeDocRef);
      return true;
    } catch (error) {
      console.error('Error deleting employee:', error);
      throw error;
    }
  }

  // Get dashboard stats
  static async getDashboardStats() {
    try {
      const employeesCollectionRef = collection(db, 'employees');

      // Total employees
      const totalQuery = query(employeesCollectionRef);

      // Active employees
      const activeQuery = query(employeesCollectionRef, where('status', '==', 'active'));

      // PT employees
      const ptQuery = query(employeesCollectionRef, where('position', '==', 'PT'));

      // Recent hires (last 30 days)
      const thirtyDaysAgo = new Date();
      thirtyDaysAgo.setDate(thirtyDaysAgo.getDate() - 30);
      
      const recentQuery = query(
        employeesCollectionRef, 
        where('startDate', '>=', Timestamp.fromDate(thirtyDaysAgo))
      );

      // Execute all queries in parallel using getCountFromServer for better performance and permissions
      const [totalSnapshot, activeSnapshot, ptSnapshot, recentSnapshot] = await Promise.all([
        getCountFromServer(totalQuery),
        getCountFromServer(activeQuery),
        getCountFromServer(ptQuery),
        getCountFromServer(recentQuery)
      ]);

      return {
        total: totalSnapshot.data().count,
        active: activeSnapshot.data().count,
        pt: ptSnapshot.data().count,
        recentHires: recentSnapshot.data().count
      };
    } catch (error) {
      console.error('Error getting employee stats:', error);
      return {
        total: 0,
        active: 0,
        pt: 0,
        recentHires: 0
      };
    }
  }

  // Update client count for PT
  static async updateClientCount(employeeId, clientCount) {
    try {
      const employeeDocRef = doc(db, 'employees', employeeId);
      await updateDoc(employeeDocRef, {
        totalClients: clientCount,
        updatedAt: Timestamp.now()
      });
      return true;
    } catch (error) {
      console.error('Error updating client count:', error);
      throw error;
    }
  }

  static async existsByIdCard(idCard, excludeId = null) {
    try {
      console.log('🔍 EmployeeModel.existsByIdCard - Checking:', { idCard, excludeId });
      
      const q = query(
        collection(db, 'employees'),
        where('idCard', '==', idCard)
      );
      
      const querySnapshot = await getDocs(q);
      
      // If no documents found, ID card doesn't exist
      if (querySnapshot.empty) {
        return false;
      }
      
      // If excludeId is provided, check if found document is different from excluded one
      if (excludeId) {
        const foundDoc = querySnapshot.docs[0];
        const foundId = foundDoc.id;
        const isDifferent = foundId !== excludeId;
        console.log('🔍 Found document:', { foundId, excludeId, isDifferent });
        return isDifferent;
      }
      
      // ID card exists
      console.log('❌ ID card exists and no excludeId provided');
      return true;
    } catch (error) {
      console.error('Error checking ID card existence:', error);
      throw error;
    }
  }
}

export default EmployeeModel;